### PR TITLE
Enhancement of FX vanilla option product, trade and pricers

### LINKED ISSUE #816: Simplify FX vanilla option
* Simplify FxVanillaOption and related classes.
* Prerequisite for exotic options.

### PR DESCRIPTION
**Pull request by yukiiwashita**
* This work closes #816

### CODE REVIEW

File: modules/product/src/main/java/com/opengamma/strata/product/fx/FxVanillaOption.java [hunk 1]
```diff
@@ -46,26 +40,17 @@
...
  * <p>
- * If the option is a call, the option holder has the right to enter into the specified exchange.
- * If the option is a put, the option holder has the right to enter into the opposite of the specified exchange.
- * <p>
  * For example, a call on a 'EUR 1.00 / USD -1.41' exchange is the option to
  * perform a foreign exchange on the expiry date, where USD 1.41 is paid to receive EUR 1.00.
- * A put on the same exchange is the option to pay EUR 1.00 and receive USD 1.41.
+ * <p>
+ * Note that all of the computation in product pricers must be based on the counter currency of the underlying FX 
+ * transaction. Thus price, pv and risk measures of the product will be expressed in USD by default for the above example.
```
Comment (marc-opengamma) [inline modules/product/src/main/java/com/opengamma/strata/product/fx/FxVanillaOption.java]: Comments about the pricer should be at the pricer level, not in the instrument description.

File: modules/product/src/main/java/com/opengamma/strata/product/fx/FxVanillaOption.java [hunk 2]
```diff
@@ -103,31 +88,13 @@
    */
   @PropertyDefinition(validate = "notNull")
```
Comment (marc-opengamma) [inline modules/product/src/main/java/com/opengamma/strata/product/fx/FxVanillaOption.java]: At expiry, the long party will have the option to enter in this transaction; the short party will, at the option of the long party, potentially enter into the inverse transaction.

File: modules/product/src/main/java/com/opengamma/strata/product/fx/ResolvedFxVanillaOption.java [hunk 3]
```diff
@@ -44,8 +40,8 @@
  * Applications will typically create a {@code ResolvedFxVanillaOption} from a {@code FxVanillaOption}
  * using {@link FxVanillaOption#resolve(ReferenceData)}.
  * <p>
- * If the option is a call, the option holder has the right to enter into the specified exchange.
- * If the option is a put, the option holder has the right to enter into the opposite of the specified exchange.
+ * Note that all of the computation in product pricers must be based on the counter currency of the underlying FX 
+ * transaction. Thus price, pv and risk measures of the product will be expressed in this currency by default.
```
Comment (marc-opengamma) [inline modules/product/src/main/java/com/opengamma/strata/product/fx/ResolvedFxVanillaOption.java]: Comments regarding pricer should be documented in the pricer.

File: modules/product/src/main/java/com/opengamma/strata/product/fx/FxVanillaOption.java [hunk 4]
```diff
@@ -46,26 +40,17 @@
...
-   * Whether the option is put or call.
-   * <p>
-   * A call gives the owner the right, but not obligation, to exercise the underlying foreign exchange.
-   * A put gives a similar option to exercise the inverse of the underlying.
-   */
-  @PropertyDefinition(validate = "notNull")
-  private final PutCall putCall;
-  /**
    * Whether the option is long or short.
    * <p>
```
Comment (marc-opengamma) [inline modules/product/src/main/java/com/opengamma/strata/product/fx/FxVanillaOption.java]: Long indicates that the option owner has the right to enter into the transaction.
Short indicates that the decision on entering the option at expiry will be exercised by the other party.

Remove the existing long/short description which is incorrect.

File: modules/product/src/main/java/com/opengamma/strata/product/fx/ResolvedFxVanillaOption.java [hunk 5]
```diff
@@ -56,14 +52,6 @@
...
-   * Whether the option is put or call.
-   * <p>
-   * A call gives the owner the right, but not obligation, to exercise the underlying foreign exchange.
-   * A put gives a similar option to exercise the inverse of the underlying.
-   */
-  @PropertyDefinition(validate = "notNull")
-  private final PutCall putCall;
-  /**
    * Whether the option is long or short.
    * <p>
```
Comment (marc-opengamma) [inline modules/product/src/main/java/com/opengamma/strata/product/fx/ResolvedFxVanillaOption.java]: Long indicates that the option owner has the right to enter into the transaction.
Short indicates that the decision on entering the option at expiry will be exercised by the other party.

Remove the existing long/short description which is incorrect.

File: modules/product/src/test/java/com/opengamma/strata/product/fx/FxVanillaOptionTest.java [hunk 6]
```diff
@@ -54,79 +54,15 @@ public void test_builder() {
     assertEquals(test.getExpiryZone(), EXPIRY_ZONE);
     assertEquals(test.getExpiryTime(), EXPIRY_TIME);
     assertEquals(test.getLongShort(), LONG);
-    assertEquals(test.getPayoffCurrency(), USD);
```
Comment (marc-opengamma) [inline modules/product/src/test/java/com/opengamma/strata/product/fx/FxVanillaOptionTest.java]: CALL and STRIKE in lines 41 and 42 are not used anymore. Delete.

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/fx/BlackFxVanillaOptionProductPricer.java [hunk 7]
```diff
@@ -74,7 +77,7 @@ public double price(
```
Comment (marc-opengamma) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/fx/BlackFxVanillaOptionProductPricer.java]: Improve price javadoc.
The price of the product is the value on the valuation date for one unit of the base currency and is expressed in the counter currency. The price does not take into account the long/short flag. See {@link #presentValue} for scaling and currency.

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/fx/BlackFxVanillaOptionProductPricer.java [hunk 8]
```diff
@@ -94,7 +97,7 @@ public CurrencyAmount presentValue(
       BlackVolatilityFxProvider volatilityProvider) {
```
Comment (marc-opengamma) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/fx/BlackFxVanillaOptionProductPricer.java]: Javadoc: The present value of the product is the value on the valuation date. It is expressed in the counter currency.

File: modules/product/src/main/java/com/opengamma/strata/product/fx/ResolvedFxVanillaOption.java [hunk 9]
```diff
@@ -123,12 +90,31 @@ public LocalDate getExpiryDate() {
...
+   * 
+   * @return the strike
+   */
+  public double getStrike() {
+    return Math.abs(underlying.getCounterCurrencyPayment().getAmount() /
+        underlying.getBaseCurrencyPayment().getAmount());
+  }
+
+  /**
+   * Gets the put/call
```
Comment (marc-opengamma) [inline modules/product/src/main/java/com/opengamma/strata/product/fx/ResolvedFxVanillaOption.java]: Returns the put/call flag. This is the put/call for the base currency. If the amount for the base currency is positive, the option is a call on the base currency (put on counter currency) and if the amount for the base currency is negative, the option is a put on the base currency (call on counter currency).

**Comment by yukiiwashita** [comment]
Closed by #882

**Pull request closed without merging**